### PR TITLE
add step_bag fixture

### PR DESCRIPTION
**Pull request by j-carson**
Creates a step_bag that makes a new pytest_harvest results_bag per step

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -503,13 +503,20 @@ You will for example obtain this kind of pivoted table:
 | test_my_app_bench[B-1] |            1 | my dataset #B | passed         |             0       |        0.870705  | passed         |            0        |
 | test_my_app_bench[B-2] |            2 | my dataset #B | passed         |             0       |        0.764746  | passed         |            1.0004   |
 
-### c- Examples
+### c- step_bag fixture
```
Comment (smarie) [inline docs/index.md:506]: nitpick

```suggestion
### c- `step_bag` fixture
```

File: docs/index.md [hunk 2]
```diff
@@ -247,7 +247,7 @@ collected 4 items
 ========================== 4 passed in 0.07 seconds ===========================
 ```
 
-#### *fixtures*
+#### <a id="fixtures"></a> *fixtures*
```
Comment (smarie) [inline docs/index.md:250]: I suggest rather to remove the italic here and on the previous subsection title (parameters).

File: docs/index.md [hunk 3]
```diff
@@ -503,13 +503,20 @@ You will for example obtain this kind of pivoted table:
 | test_my_app_bench[B-1] |            1 | my dataset #B | passed         |             0       |        0.870705  | passed         |            0        |
 | test_my_app_bench[B-2] |            2 | my dataset #B | passed         |             0       |        0.764746  | passed         |            1.0004   |
 
-### c- Examples
+### c- step_bag fixture
 
-Two examples are available that should be quite straightforward for those familiar with pytest-harvest:
+As explained [above](#fixtures), the behavior of function level fixtures 
+can be unexpected when combined with steps. The `step_bag` fixture is a
+version of the pytest-harvest `results_bag` fixture, but decorated with
```
Comment (smarie) [inline docs/index.md:510]: A link to pytest harvest would be nice:

```suggestion
version of the pytest-harvest [`results_bag`](https://smarie.github.io/python-pytest-harvest/#b-collecting-test-artifacts) fixture, but decorated with 
```

**Comment by smarie** [comment]
Thanks @j-carson  ! very nice PR. I made a few minor comments

Unfortunately I did not yet migrate this project to my new template with github actions :( So I'll have to update it first

File: docs/index.md (hunk 2 above)
Comment (j-carson) [reply docs/index.md:250]: Not quite sure what you mean here?  I created the anchor link so I could reference this area further down (line 508). Even if it's not in italics, I would still want to declare an anchor? Or does mkdocs make anchors for me? (I'm used to sphinx)

**Comment by j-carson** [comment]
I will update the PR once I understand what you meant on that one doc comment. I obviously noticed there were build issues - I put what I did to make the repo build in a separate PR so you could see what I did to test this PR while keeping it separate from the code that was more likely to be actually merged.

-- 
  Janet Carson
  ***@***.***

File: docs/index.md (hunk 2 above)
Comment (smarie) [reply docs/index.md:250]: mkdocs does it automatically

File: pytest_steps/plugin.py [hunk 4]
```diff
@@ -30,3 +31,11 @@ def module_results_df_steps_pivoted(request, module_results_df):
...
         # Pivot
         return pivot_steps_on_df(module_results_df, pytest_session=request.session)
+
+    @pytest.fixture
+    @one_fixture_per_step
+    def step_bag(results_bag):
+        """
+        Provides a separate pytest-harvest "results_bag" per step
+        """
+        return results_bag
```
Comment (smarie) [inline pytest_steps/plugin.py:41]:
```suggestion
        return results_bag

```

File: pytest_steps/tests/test_steps_harvest_step_bag.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    yield
+
+
+# Run the above silly test with each type of bag
+
+
+@test_steps("apples", "bananas")
+def test_fruit_step_bag(step_bag):
+    # This test will have a new bag per step
+    yield from fruit_counter(step_bag)
```
Comment (smarie) [inline pytest_steps/tests/test_steps_harvest_step_bag.py:30]:
```suggestion
    for f in fruit_counter(step_bag):
        yield f
```

File: pytest_steps/tests/test_steps_harvest_step_bag.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+
+@test_steps("apples", "bananas")
+def test_fruit_results_bag(results_bag):
+    # This test will have a shared bag over both steps
+    yield from fruit_counter(results_bag)
```
Comment (smarie) [inline pytest_steps/tests/test_steps_harvest_step_bag.py:36]:
```suggestion
    for f in fruit_counter(step_bag):
        yield f
```

**Comment by j-carson** [comment]
As I was reworking the docs I realized that to be logically complete, in addtion  to providing a fixture that overrides the default results_bag for generator-mode steps, I should provide the complementary cross-steps bag for explicit-mode steps.

**Comment by smarie** [comment]
Apart from the failing tests in legacy python versions (see my comments) it seems ok. Nice suggestion to also provide the cross-steps bag!

**Comment by smarie** [comment]
Apparently a few jobs are still failing. I'll have a look tomorrow

**Comment by j-carson** [comment]
I think I fixed it.

**Review by smarie — APPROVED** [review]

**Pull request merged**